QuickEdit: Add Featured Image Control

gigitux opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Part of #64519 

This PR adds the Featured Image Control for the Page view. 



https://github.com/user-attachments/assets/cc8d5a58-97fd-4ac5-b1cc-ed705f669bd4



## View mode

| Grid | Table | Layout |
|--------|--------|--------|
| ![image](https://github.com/user-attachments/assets/2d8820a6-61e7-4bcf-ada5-89170c4f960a) | ![image](https://github.com/user-attachments/assets/7ab4799d-8439-4a8f-ab9c-6e1179413583) | ![image](https://github.com/user-attachments/assets/7634d2bb-9823-4133-be9f-8b009ad3e6f5) | 


## Edit

| Without Image | With Image |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/ae8c22cd-b55c-49e9-b953-3fea9e153ec1) | ![image](https://github.com/user-attachments/assets/222358eb-14b8-42ea-b2ef-dd332d402d02) | 


## Additional Changes

### Render component and view prop

With this PR, the render component will now receive the view prop. This addition is necessary because the component’s layout needs to adapt based on the view. Similar cases include the Title field, where different layouts are required depending on the view.

https://github.com/WordPress/gutenberg/blob/83569778dba9dd64dbd6db2e4623e36457b7f7ac/packages/edit-site/src/components/post-fields/index.js#L163-L164

### Fix error on bulk edit mode

https://github.com/user-attachments/assets/c906cb68-9328-4cca-b61d-f24abf1f8861

```
Uncaught TypeError: Cannot read properties of null (reading 'status')
```

Currently, there is an error when the user edits a field in bulk mode. [This PR fixes this error](https://github.com/WordPress/gutenberg/pull/64496/files#diff-4ed486e973856f952d2b80aea381fc1f5a1d0004995a0b68554147b02cee0b55R75).

### Attachment type

This PR updates the `Attachment` type.





<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Ensure that `Quick Edit in DataViews` and `Quick Edit in DataViews` are enabled.

1. Visit the Site Editor
2. Click on pages.
3. Toggle to the table view.
4. Click on the settings.
5. Make the feature image field visible.
6. Click on details panel icon.
7. Ensure that the Featured Image option is available.
8. Click on it.
9. Ensure that the Media Gallery opens and click on an image.
10. Save.
11. Ensure that the image is set as the featured image to the post.
12. Toggle to the grid view.
13. Ensure that the image is rendered correctly.
14. Toggle to the list view.
15. Ensure that the image is rendered correctly.



### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

gigitux committed:
Page Edit View: Implement Featured image page field

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

jameskoster commented:
To be clear, I think the Featured Image control should remain as it is currently in terms of design and position. 

However I also recognise that panel Data Forms should be able to handle `image` fields, respecting the established conventions of this layout. Here's a mockup and flow that could achieve that:

<img width="2934" alt="image" src="https://github.com/user-attachments/assets/e59c1283-6b24-4208-b3c3-d036514fce96">

What do you think?

gigitux commented:
> To be clear, I think the Featured Image control should remain as it is currently in terms of design and position.
> 
> However I also recognise that panel Data Forms should be able to handle `image` fields, respecting the established conventions of this layout. Here's a mockup and flow that could achieve that:
> 
> <img alt="image" width="2000" src="https://private-user-images.githubusercontent.com/846565/358244757-e59c1283-6b24-4208-b3c3-d036514fce96.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDIzMzgsIm5iZiI6MTcyMzgwMjAzOCwicGF0aCI6Ii84NDY1NjUvMzU4MjQ0NzU3LWU1OWMxMjgzLTZiMjQtNDIwOC1iM2MzLWQwMzY1MTRmY2U5Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxNlQwOTUzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Y2JjMTY5NGZjMTg1ZDlmZjY2NzUzNzZhYTQyMDkzMDI2MTdmNjNkZDQxZWFmNTY1M2JkZWI4YWZhNjU1NDFmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.IP_nt_JOYwHQy5tn_r5cGcEDoCUveJSqtDtG7_A9x6Q">
> What do you think?

Thanks for your feedback! It makes sense! I'm going to implement your feedback! 🚀

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
improve flow and style

gigitux commented:
Hey @jameskoster, I've implemented your suggestion: 


https://github.com/user-attachments/assets/bb11cc5f-4ea4-4693-ae98-420da4ed16aa

Let me know if, from the design perspective, there is something to address.


I’m not marking this PR as ready for review because we still need to implement the default edit control for the image field. Since DataViews is a general-purpose library, we shouldn’t rely on the WordPress Media Gallery for the generic control. Instead, I suggest using the Operating System’s file selection option:

https://github.com/user-attachments/assets/b73f3de8-3d87-4cdd-87d9-e94c953f9fde

However, please note that in this case, we won’t have access to the filename once the file is uploaded.

gigitux committed:
fix list layout

jameskoster commented:
Thanks for the update. The flow seems pretty good, but there are some design details to fix.

<img width="310" alt="Screenshot 2024-08-22 at 16 06 34" src="https://github.com/user-attachments/assets/336ae3d8-1e6c-44e2-a12e-77f9ff512567">


* Border radius is missing
* Border should use `$border-width` (not the focus width)
* Border color should be `$gray-300`
* Hover background should be `$gray-100`
* Thumbnail should be `24px`
* The thumbnail appearance should match unset color swatches
* Gap between swatch and text should be `8px` (seems you can remove the margin on the `span`)
* Padding should be `8px 12px`

Also here:

<img width="328" alt="Screenshot 2024-08-22 at 16 09 13" src="https://github.com/user-attachments/assets/b467e014-ae7d-43f6-b472-d7eebb480b05">

* Thumbnail should be `24px` with `$radius-small` border-radius. Ideally there should be a semi-transparent black outline so that white images are still visible. You can see an example in data views if you toggle on the preview.
* Image name container height should be `24px` to align with thumbnail
* Filename should be `12px`, with `16px` line-height
* Gap between thumbnail and text should be `8px` (again, removing the margin on the span should fix)
* Padding should be `8px 12px`
* Reset button size should be `small` (`24px`) to align with the image name / thumbnail
* Ideally the filename would be truncated so that you can see the filetype. Perhaps this can be accomplished by using the [`Text`](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-text--docs) component with the `truncate` and `ellipsizeMode ` props.

Here's a mockup that hopefully clarifies some of the details:

<img width="854" alt="Screenshot 2024-08-22 at 16 20 23" src="https://github.com/user-attachments/assets/8c0ba262-817a-4008-bcc4-469c21000d07">

gigitux committed:
improve design

gigitux commented:
With 9a39fe6, I addressed your feedback!


https://github.com/user-attachments/assets/0966a669-a3af-4b9d-82dd-3eb15f4cca2c

The only remaining challenge is that the filename can only be truncated to a fixed number of characters, which may result in it not always ending directly under the remove button.

gigitux commented on packages/edit-site/src/components/post-fields/index.js:
```diff
@@ -395,8 +362,212 @@ function usePostFields( viewType ) {
+												tabIndex={ 0 }
+												onClick={ ( event ) => {
+													const element =
+														event.target.tagName.toLowerCase();
+													// Prevent opening the media modal when clicking on the button/icon.
+													if (
+														element !== 'button' &&
+														element !== 'svg'
+													) {
```
The issue is that there's a button within a clickable area, and clicking the button triggers a separate action. I'm not sure if this is the best way to handle it.

jameskoster replied:
Yeah we should avoid nested buttons. We had a similar issue in data views list layout... this UI:

<img width="384" alt="Screenshot 2024-08-23 at 11 41 02" src="https://github.com/user-attachments/assets/11cea8e7-515d-423e-a8fe-059d983f0311">

Perhaps some of the techniques there can be borrowed?

gigitux commented on packages/edit-site/src/components/post-fields/index.js:
```diff
@@ -395,8 +362,212 @@ function usePostFields( viewType ) {
+									render={ ( { open } ) => {
+										return (
+											<div
+												role="button"
```
Related comment: https://github.com/WordPress/gutenberg/pull/64496/files#r1728584779

gigitux committed:
remove not necessary style

jameskoster commented:
Nice, it's getting there :)

Border-radius on the thumbnail and button are still missing.

Truncation of the file name seems to occur slightly before it's necessary. In the screenshot below note that there's more space the title could occypy.

<img width="309" alt="Screenshot 2024-08-23 at 11 45 55" src="https://github.com/user-attachments/assets/bfe73c0f-38b1-4317-b514-ffc23b1ff3ea">

I know the button that appears in the panel is a work in progress, but I wonder if it could have the same appearance as the author field. What do you think?

<img width="342" alt="Screenshot 2024-08-23 at 11 49 09" src="https://github.com/user-attachments/assets/15b5f0f9-6d95-4549-a244-31c3585f9f77">

gigitux commented:
> Truncation of the file name seems to occur slightly before it's necessary. In the screenshot below note that there's more space the title could occypy.

As I mentioned in the previous comment, this could be complex to achieve given that `Text` components can only truncate to a fixed number of characters, which may result in it not always ending directly under the remove button.

jameskoster commented:
> given that Text components can only truncate to a fixed number of characters

Are you sure? It seems to truncate based on width when `truncate` is `true` and `numberOfLines` is `1`?

> Do you mean something like this?

Sorry I should have been more explicit. See the mockup below and notice that the Author and Featured Image buttons (outlined in red) share the same characteristics; thumbnail size, gap, spacing, etc.

<img width="580" alt="Screenshot 2024-08-23 at 12 23 12" src="https://github.com/user-attachments/assets/0fb08a78-1c21-4d06-81f7-ef0f202bf1a0">

gigitux commented:
> Are you sure? It seems to truncate based on width when `truncate` is `true` and `numberOfLines` is `1`?
> 
> >

[The documentation](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-text--docs) says that when `ellipsizeMode` is `middle` (our case), the limit property is required:

<img width="956" alt="image" src="https://github.com/user-attachments/assets/a3cf076f-5268-4f1c-af5a-f35582f936fb">


´Limit` is the max number of characters to be displayed:
<img width="995" alt="image" src="https://github.com/user-attachments/assets/9ea7a56f-4b16-4afa-bb36-dbc002793721">

So, if we want to truncate in the middle (so, showing the extension), we need to explicit the max number of characters.

> Sorry I should have been more explicit. See the mockup below and notice that the Author and Featured Image buttons (outlined in red) share the same characteristics; thumbnail size, gap, spacing, etc.

Gothca! Thanks! 🙇

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
improve style

gigitux committed:
improve default image control

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

ntsekouras commented on packages/edit-site/src/components/post-fields/index.js:
```diff
@@ -179,26 +150,22 @@ function usePostFields( viewType ) {
-				label: __( 'Featured Image' ),
-				getValue: ( { item } ) => item.featured_media,
-				render: ( { item } ) => (
-					<FeaturedImage item={ item } viewType={ viewType } />
```
Two questions:
1. Was there a reason to change the order here?
2. What was the reason for inlining the `render` function instead of keeping the `FeaturedImage` component?

ntsekouras commented on packages/edit-site/src/components/sidebar-dataviews/default-views.js:
```diff
@@ -31,7 +31,7 @@ export const defaultLayouts = {
 			primaryField: 'title',
 			styles: {
-				'featured-image': {
+				featured_media: {
```
Why this was needed?

ntsekouras commented:
Thanks for the PR! I quickly tried this with `table` layout and the display doesn't seem good:


https://github.com/user-attachments/assets/79926d11-753b-4c9d-8138-11303e41534a

Is this something that you can also see?

gigitux committed:
fix style

gigitux committed:
remove not necessary configuration

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
format _z-index.scss file

gigitux committed:
improve style

gigitux committed:
remove not necessary code

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux commented:
Thanks @ntsekouras for your precious feedback! I addressed your suggestion! Code-wise the PR isn't ready to be reviewed, mostly because I'm going to explore the solution that you suggested [here](https://wordpress.slack.com/archives/C02QB2JS7/p1724830339489289?thread_ts=1724748812.359799&cid=C02QB2JS7): passing mediaUpload to the generic image field control.

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
fix focus

gigitux committed:
remove image field type

gigitux committed:
add comment

gigitux commented:
I'm going to mark this PR as ready to be reviewed. I'm going to update the description, but overall this PR introduces the Featured Image control following the @jameskoster feedback.

The remaining task is to truncate the filename based on the available width. Currently, truncation is performed using a fixed number of characters. Is it something that, is it needed? 

In the future, this code may be moved to the @wordpress/core-fields-control package if we agree to create it.

gigitux committed:
fix warning

gigitux committed:
fix image for deleted pages

jameskoster commented:
> The remaining task is to truncate the filename based on the available width. Currently, truncation is performed using a fixed number of characters. Is it something that, is it needed?

If this is too problematic we could probably omit the filename for now.

Look at the screenshot in your last comment it seems the radius is missing on the thumbnail and the container. We can use `$radius-small` for this.

gigitux committed:
remove filename

gigitux committed:
add border-radius

oandregal commented on packages/edit-site/src/components/post-fields/index.js:
```diff
@@ -191,12 +233,117 @@ function usePostFields( viewType ) {
+				id: 'featured_media',
 				label: __( 'Featured Image' ),
-				getValue: ( { item } ) => item.featured_media,
+				type: 'image',
```
What is this `type` for? This PR provides custom `Edit` and `render` functions and the `type` is not used anywhere I can see.

I see a few options:

1. Implement the `image` type for DataForm (with a demo in storybook), like [any other types](https://github.com/WordPress/gutenberg/blob/trunk/packages/dataviews/src/dataform-controls/index.tsx#L1)) we support. The advantage is that we can make `type` a required prop to fields. Then, we'd override the Edit control here to use WordPress-specifics.
2. Implement a custom `feature-image` type that is specific to WordPress, like the [slug PR](https://github.com/WordPress/gutenberg/pull/65196) tries to do.
3. Do not provide any type and just override Edit & render. We'd still need to 1 at some point. It's fine if it's a follow-up.

gigitux replied:
This PR has been opened before `@wordpress/fields` package and the multiple great conversations that we had during these week. I think that it makes sense to follow the second option. I'm going to move this PR as draft, given that it will require some work before another round of review!

Thanks for the help! 🙇

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
migrate featured image to fields package

gigitux committed:
fix CSS

gigitux committed:
remove not necessary changes

gigitux commented on packages/fields/src/fields/featured-image/featured-image-edit.tsx:
```diff
@@ -0,0 +1,122 @@
+	const ref = useRef( null );
+
+	return (
+		<fieldset className="fields-controls__featured-image">
```
What do you think about this class name? I asked a similiar question in this PR: https://github.com/WordPress/gutenberg/pull/65196#discussion_r1764760161

gigitux committed:
fix type

gigitux committed:
remove empty space

gigitux commented:
I'm going to mark this PR as ready to be reviewed!

gigitux committed:
fix z-index gallery

gigitux committed:
fix overlapping and style

gigitux committed:
improve codestyle

gigitux commented:
Thanks for catching those!

> There's a z-index overlap with the media library:
> Long names are overflowing, should we truncate?

Both issue are addressed:

https://github.com/user-attachments/assets/d4a6191e-32bb-4c76-84db-c490489087c0

For the long names in the panel, we have the issue with all the text fields:

![image](https://github.com/user-attachments/assets/1ff24c90-3b2f-4e9d-b148-35fe40a6c13a)

If you agree, I will open an issue and I will fix it in a dedicated PR given that the fix will impact changes to the DataView library.

> Featured images in data views List layout are being cropped:

I'm unable to replicate it! Technically, in this view, the data view library CSS should be applied: can you share the image?

> Featured image in table layout is smaller than trunk:

I fixed it! Do you think we should use the same placeholder that’s used in the panel view?

![image](https://github.com/user-attachments/assets/416913d0-c095-4fc9-8039-ba4b28c6c01b)

jameskoster commented:
> If you agree, I will open an issue and I will fix it in a dedicated PR given that the fix will impact changes to the DataView library.

Yep, good idea.

> I fixed it! Do you think we should use the same placeholder that’s used in the panel?

Let's stick with the gray box for now, to keep visual changes to a minimum and maintain alignment with grid/list layouts.

> I'm unable to replicate it! Technically, in this view, the data view library CSS should be applied: can you share the image?

Never mind, it seems to be an issue on trunk too. Strangely enabling the 'posts dashboard' experiment fixes. We can fix this separately.

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
use same placeholder style grid/list layout

gigitux commented:
> Let's stick with the gray box for now, to keep visual changes to a minimum and maintain alignment with grid/list layouts.

Fixed with 1b84ac0. Thanks for your feedback and patience! 🙇 

![image](https://github.com/user-attachments/assets/da06eb57-5452-4d67-ad95-82a2bae1b9a1)

> Yep, good idea.

👍 

> Never mind, it seems to be an issue on trunk too. Strangely enabling the 'posts dashboard' experiment fixes. We can fix this separately.

👍

gigitux committed:
fix tsconfig

jameskoster commented:
> Never mind, it seems to be an issue on trunk too. Strangely enabling the 'posts dashboard' experiment fixes. We can fix this separately.

On closer inspection I may have been wrong about that. You should be able to replicate if you use a landscape image as a featured image. 

I made a PR to fix this on trunk (https://github.com/WordPress/gutenberg/pull/65525), but we should probably apply it here instead.

oandregal commented on packages/edit-site/src/components/sidebar-dataviews/default-views.js:
```diff
@@ -30,10 +30,8 @@ export const defaultLayouts = {
 	[ LAYOUT_TABLE ]: {
 		layout: {
 			primaryField: 'title',
+			mediaField: 'featured_media',
```
List & Grid have "media fields", meaning there are special spots for rendering this field. What does it mean the "media field" of the table?

gigitux replied:
I removed it with 2b595c20f63b245b7fbbbe27a512ac11e6079701

oandregal commented on packages/fields/src/fields/featured-image/index.ts:
```diff
@@ -0,0 +1,23 @@
+import { FeaturedImageView } from './featured-image-view';
+
+const featuredImageField: Field< BasePost > = {
+	id: 'featured_media',
```
What's the `type` for this field? We should make progress on making the `type` obligatory for fields and every new field we add should introduce the corresponding type if there's one. Is it `attachment`, `image`, etc.?

oandregal replied:
An additional thought: I think we'd want to have two behaviors for images:

- an attachment, like in this PR
- an embedded image, like the feature image is in the document inspector

How that would look like?

gigitux replied:
To clarify, are you suggesting adding the type of entity that the field will display or edit?

If so, I think that we need to discuss this further. After some chat with @louwie17, the idea is the type field should be specific to only one field component.

In the future, the type field will be essential for registering the dedicated field in the Redux store, establishing a unique 1:1 relationship between the type and the field.

To elaborate further, we are planning to develop the `registerFieldType` API, which will be essential for registering new custom fields. The type value will distinguish one custom field from another.

With c448c494b25d9bf31b5e3c0b76c07e2f9a88bd04, I added `featured-image` as type.

In any case, I'm happy to discuss this further. It is really important that we are aligned on this topic! 🙇

gigitux replied:
We are discussing the same topic [here](https://github.com/WordPress/gutenberg/pull/65196#discussion_r1782971224)

youknowriad replied:
I think the "type" here should be "media" and ideally we wouldn't need to define an "edit" function in that case. All media in WordPress could use that type. That said, that's dependent on the possibility to register field types which we don't have at the moment. So I'd suggest for this PR staying with something low level like "number" or "string".

gigitux replied:
Fixed with 9cc6d80958c9a240524cbe0e715b81a38030e37a

youknowriad replied:
I actually suggested "integer" or "text" instead of "media" for now. I don't think we should be using a type that doesn't have any registered definition.

gigitux replied:
Ops, sorry! With 01a57a861348a49d641b7638be365442efbca97a should be fine now!

oandregal commented on packages/dataviews/src/dataforms-layouts/panel/index.tsx:
```diff
@@ -111,7 +112,7 @@ function FormField< Item >( {
 							onClick={ onToggle }
 						>
-							<field.render item={ data } />
+							<field.render item={ data } view={ LAYOUT_PANEL } />
```
How this looks like in the regular panel? Can we update the storybook to see how fields of this type behave in both layouts?

gigitux replied:
This change does not affect the appearance of the current fields in different views. The primary goal is to pass this information to the field, enabling it to adjust its layout based on the view property.

oandregal commented on packages/fields/src/fields/featured-image/featured-image-view.tsx:
```diff
@@ -0,0 +1,176 @@
+	url: string | undefined;
+	onClick: () => void;
+} ) => ( {
+	grid: {
```
How this component will be displayed when the view type is unknown? We aim to allow 3rd parties to register/unregister layouts for the DataViews.

Can we make it so this render function is independent from the view type? Thinking out loud: perhaps this render function always return an `img` and the layout can add wrappers based on the field type.

jameskoster commented:
Hey @gigitux are you still working on this one?

jameskoster commented:
I ask because out of conversations in other issues came a realisation that we might improve the design by making the 'Choose image..." button a drop zone:

<img width="1537" alt="image" src="https://github.com/user-attachments/assets/0bb486fb-7e96-42c5-b91a-5db3b10abfcc">

gigitux commented:
Hey @jameskoster, thanks for the ping!
Currently, we are on cooldown, so I'm prioritizing other tasks (not related to the data views). During the upcoming weeks, we will start to work on this project and I will take care of updating this field.


> I ask because out of conversations in other issues came a realisation that we might improve the design by making the 'Choose image..." button a drop zone:

Just to clarify, if the user clicks on the `Choose Image` button, should we still render the dropdown menu?

![image](https://github.com/user-attachments/assets/909232ef-78bb-4f4e-9a5a-d6688e677f79)

jameskoster commented:
> Just to clarify, if the user clicks on the Choose Image button, should we still render the dropdown menu?

I'm curious to hear your thoughts on that from the technical perspective. Is it a requirement of this `type` in the Data Forms component that all fields are configured in a popover? For most fields it's okay, but for this particular one (images) it adds an unnecessary step in the UX that I think would be good to avoid.

gigitux commented:
> > Just to clarify, if the user clicks on the Choose Image button, should we still render the dropdown menu?
> 
> I'm curious to hear your thoughts on that from the technical perspective. Is it a requirement of this `type` in the Data Forms component that all fields are configured in a popover? For most fields it's okay, but for this particular one (images) it adds an unnecessary step in the UX that I think would be good to avoid.

Currently, it is needed, but @louwie17 shared with me [this discussion](https://github.com/WordPress/gutenberg/pull/65399#discussion_r1777452478).


It is something that we can't implement ASAP, but we are planning to investigate how consumers can control the layout. We will take in consideration this use-case during our exploration! Thanks for raising this up! 🙇 

cc @oandregal @youknowriad

youknowriad commented:
I think it would be a great idea to explore removing the "layout" type from the form. This use-case (mix and match regular form elements with drop-down) kind of confirms it.

youknowriad commented:
That said, I think this PR is going in a good direction, I don't think we should block it because of that. I'm sure @jameskoster can agree with me that it should be fine to land this as a dropdown and we figure out the removal of layout separately.

jameskoster commented:
Yup, totally!

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
add type

gigitux committed:
remove mediaField from layout table

gigitux commented:
> That said, I think this PR is going in a good direction, I don't think we should block it because of that. I'm sure @jameskoster can agree with me that it should be fine to land this as a dropdown and we figure out the removal of layout separately.

I replied to some comments. To clarify, based on this comment, it looks like that the only blocker to merge this PR is [this discussion](https://github.com/WordPress/gutenberg/pull/64496#discussion_r1777554179), right? 

Instead, in the backlog for the future, we are planning to:
- new UI with drop zone
- removing the "layout" type from the form (this is fundamental when we will plan to allow 3rd parties to register/unregister layouts)

If you can confirm, I will take care to open dedicated issues for these tasks.

youknowriad commented:
Yes, I don't see much blockers here. I replied to the "type" discussion.

youknowriad commented on packages/dataviews/src/dataviews-layouts/grid/index.tsx:
```diff
@@ -54,10 +55,10 @@ function GridItem< Item >( {
 	const isSelected = selection.includes( id );
 	const renderedMediaField = mediaField?.render ? (
-		<mediaField.render item={ item } />
+		<mediaField.render item={ item } view={ LAYOUT_GRID } />
```
Why adding a view to the render function is necessary? I'd like us to avoid this if possible.

gigitux replied:
Based on the layout, the field has a different UI:

https://github.com/WordPress/gutenberg/blob/9cc6d80958c9a240524cbe0e715b81a38030e37a/packages/fields/src/fields/featured-image/featured-image-view.tsx#L167-L169

youknowriad replied:
Why?

youknowriad replied:
Let me clarify:

People are going to be able to register data views layouts independently from fields, meaning ideally the rendering of fields is universal and shouldn't be dependent on the layout type.

gigitux replied:
While I favor the universal approach, I'm afraid that it will not be very flexible. In the case of the featured image field, we have major differents between panel and the rest of the layouts:

| Type | Image | Without Image|
|--------|--------|--------|
| Panel | ![image](https://github.com/user-attachments/assets/812aa865-633c-4a0c-96d2-8f3fd625135d) | ![image](https://github.com/user-attachments/assets/1e470fce-de3f-4b76-8f2a-01ff5d9725a6)|
| Grid | ![image](https://github.com/user-attachments/assets/a07e6129-0c08-46d2-ac97-f8b94d3af5df) | ![image](https://github.com/user-attachments/assets/041198f5-a40e-4341-9846-521b1eef6c94)
| Table  | ![image](https://github.com/user-attachments/assets/92913fdb-f25b-4b0d-829a-9cac0427e19a) | ![image](https://github.com/user-attachments/assets/ba18dfa3-74a4-43d6-892b-e423dac408fb) | the image needs to be 32px |
| List | ![image](https://github.com/user-attachments/assets/b2908a68-430e-4458-9202-73285efeca6c) | ![image](https://github.com/user-attachments/assets/b1cf42fd-3c5c-4767-878c-7aad6d4c300b) | the image needs to be  |

gigitux replied:
We will have a similar issue even [with the new layout](https://github.com/WordPress/gutenberg/pull/64496#issuecomment-2389222779).

youknowriad replied:
I know it's a difficult constraint, but I think we should think of this constraint as a good thing and instead try to solve it in a way that doesn't require knowledge about the layout.

There seem to be one main difference:

 - In a form, in addition to the thumbnail, we show either "logo-1.jpg" or "Choose an image". 
 - In dataviews, it's always just a thumbnail.

So what does this difference mean for the "field" abstraction.

 - First, I would argue that we can start (potentially merge this PR) by using just the thumbnail in the form like all the other usage in dataviews.
 - Second, if we do want to make this discussion, I think I'd ask myself the question: What is the prop that would make sense for a field to define this distinction. A prop that new layouts could decide to use or not. For instance something like `size` could make sense or something like `isDetailed` ... And then a layout or a DataForm, could decide to say, I render all the fields with a "small" size, or I render all the fields with "details", from there a Field could decide what it means for them to change size or to show details. A field might decide that it has no impact on them.

gigitux replied:
> First, I would argue that we can start (potentially merge this PR) by using just the thumbnail in the form like all the other usage in dataviews.

I'm going to proceed with this approach, given that the second one looks like not strictly related to this field control, and it needs further discussion.

gigitux replied:
Updated the PR with commit c2994e0ab6850e23b284d1a82ad7b4608b13311d to no longer rely on the view prop. The main downside of this approach is that the click event is no longer configured for the grid view. So when the user clicks on the image, the editor isn't opened anymore:

| Before | After |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/794765e8-8449-4211-9eac-fbeb7dadc0a4" /> | <video src="https://github.com/user-attachments/assets/4dfea3ba-1735-47d5-975c-7a6c54e001f8" /> |

Additionally, even though the field no longer relies on the view prop, setting the correct image dimensions still requires using the CSS cascade:

https://github.com/WordPress/gutenberg/blob/c2994e0ab6850e23b284d1a82ad7b4608b13311d/packages/fields/src/fields/featured-image/style.scss#L73-L96

gigitux replied:
> Second, if we do want to make this discussion, I think I'd ask myself the question: What is the prop that would make sense for a field to define this distinction. A prop that new layouts could decide to use or not. For instance something like size could make sense or something like isDetailed ... And then a layout or a DataForm, could decide to say, I render all the fields with a "small" size, or I render all the fields with "details", from there a Field could decide what it means for them to change size or to show details. A field might decide that it has no impact on them.

I created a dedicated issue for this topic: https://github.com/WordPress/gutenberg/issues/66336

gigitux committed:
update type

youknowriad commented on packages/fields/src/fields/featured-image/featured-image-edit.tsx:
```diff
@@ -0,0 +1,122 @@
+import { useSelect } from '@wordpress/data';
+import { useCallback, useRef } from '@wordpress/element';
+// @ts-ignore
+import { MediaUpload } from '@wordpress/block-editor';
```
I'm not sure I like the block-editor" dependency in the fields package. I guess it's because it's using a MediaUpload component that is specific to WordPress. Can we use `import { MediaUpload } from '@wordpress/media-utils';` instead ?

gigitux replied:
TIL! Nice! Fixed with 7b1e40a6f0a56a990c97e45f1c57ffa12ba899d5.

gigitux committed:
use text as type

gigitux committed:
use media-utils package

youknowriad commented on packages/fields/tsconfig.json:
```diff
@@ -23,7 +23,9 @@
 		{ "path": "../hooks" },
-		{ "path": "../html-entities" }
+		{ "path": "../html-entities" },
+		{ "path": "../block-editor" },
```
I think this is a left over (need to be replaced by media-utils maybe)

gigitux replied:
Good catch! Fixed with [6592d36](https://github.com/WordPress/gutenberg/pull/64496/commits/6592d3652ad22d751038fd9bf375ea2da5e9dafa)

gigitux committed:
fix tsconfig

gigitux committed:
remove view prop

youknowriad commented on packages/edit-site/src/components/post-fields/index.js:
```diff
@@ -190,15 +149,7 @@ function usePostFields( viewType ) {
-				),
-				enableSorting: false,
-			},
+			featuredImageField,
```
I like that we're moving the dataviews fields to the fields package one by one 👍 and fixing them (removing dependencies...)

youknowriad commented on packages/edit-site/src/components/sidebar-dataviews/default-views.js:
```diff
@@ -31,9 +31,6 @@ export const defaultLayouts = {
 			styles: {
-				'featured-image': {
-					width: '1%',
-				},
```
Was this unnecessary?

gigitux replied:
Yes!

youknowriad commented on packages/fields/src/actions/utils.ts:
```diff
@@ -34,10 +34,10 @@ export function getItemTitle( item: Post ) {
 		return decodeEntities( item.title );
 	}
-	if ( 'rendered' in item.title ) {
+	if ( item.title && 'rendered' in item.title ) {
```
In which case do we call this function with a non existent item.title ?

gigitux replied:
I remember encountering some errors in the past, which led me to add this check. However, I’m no longer able to replicate the issue. I’m happy to revert the changes.

gigitux replied:
Reverted the changes with b039310229a9c52252615070a8462bbb36b3b4f5.

youknowriad reviewed (approved):
Nice work here. 

I know the constraint of "no layout" in render is creating a non ideal result in the list view but hopefully we can address later with a better solution.

youknowriad commented:
(looks like you need to run.`npm install` and commit the lock)

youknowriad commented:
> The main downside of this approach is that the click event is no longer configured for the grid view.

I think we should follow-up by having a special callback `selectItem` similar to `getItemId` in dataViews, that layouts can trigger when they want: "grid view" can trigger it when clicking the media and primary field, table view can trigger it when clicking primary fields...

gigitux committed:
revert changes

gigitux committed:
generate package-lock.json

gigitux commented:
> (looks like you need to run.`npm install` and commit the lock)

Fixed with 42f90f4e50aab6f8b837a4d01e4c6c6acb4f517e.

youknowriad commented:
> I think we should follow-up by having a special callback selectItem similar to getItemId in dataViews, that layouts can trigger when they want: "grid view" can trigger it when clicking the media and primary field, table view can trigger it when clicking primary fields...

@gigitux Do you think this one is something you can help with? It feels like the most important follow-up here.

gigitux commented:
Thanks for your review! Based on our discussion, I created three issues:
- #66332 
- #66336 
- #66338 

> @gigitux Do you think this one is something you can help with? It feels like the most important follow-up here.

Yes, I will work on #66338!

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into add/page-image-control

gigitux committed:
generate package-lock.json

oandregal commented on packages/fields/src/fields/featured-image/style.scss:
```diff
@@ -0,0 +1,95 @@
+	}
+}
+
+.dataforms-layouts-panel__field-control {
```
How do absorb this for any field of type media so consumers don't have to use dataviews classes (they're not public API). Would it be resolved if we add a dataviews type `media` that absorbs these styles and the "featured media field" uses?

oandregal commented on packages/fields/src/fields/featured-image/featured-image-view.tsx:
```diff
@@ -0,0 +1,38 @@
+
+export const FeaturedImageView = ( {
+	item,
+}: DataViewRenderFieldProps< BasePost > ) => {
```
It seems this could be implemented in a way that is not specific to "featured media", but rather to a `media` type. For example, what if we do something along these lines:

```js
FeaturedImageView = ({
  item,
}) {
  const url = item.getValue();
  return (<img src={ url } );
}
```

If this was a `media` type field, the field "featured media" could be declared as:

```js
{
  id: 'featured-media',
  type: 'media',
  getValue: () => { /* convert item to URL here */ },
}
```

This means that field types would also provide a `render` default. I think we should assume field types should provide default for every field property.

oandregal commented:
What are the next steps to consolidate these two experiences for media?

| Editor | Quick Edit |
| --- | --- |
| <img width="282" alt="Screenshot 2024-10-23 at 18 16 36" src="https://github.com/user-attachments/assets/9a54e77d-1cf3-4e3f-a908-29fb41d4d15e"> | <img width="374" alt="Screenshot 2024-10-23 at 18 16 46" src="https://github.com/user-attachments/assets/d44abbe1-ffca-43f4-97ad-9b60e25fd7a5"> |

youknowriad commented:
@oandregal yeah, my main idea there (not sure if it's tracked) is to remove the "layout" type in the "form" config and instead have a "form.fields" config that is like that:

```
[ 
   { field: 'featured-image', label: 'hidden', control: 'inline' }, { field: 'slug', label: 'side', control: 'dropdown' } 
]
```

or something like that.

gigitux commented:
> @oandregal yeah, my main idea there (not sure if it's tracked) is to remove the "layout" type in the "form" config and instead have a "form.fields" config that is like that:
> 
> ```
> [ 
>    { field: 'featured-image', label: 'hidden', control: 'inline' }, { field: 'slug', label: 'side', control: 'dropdown' } 
> ]
> ```
> 
> or something like that.

This is something that we are planning to work on it soon... cc @louwie17
